### PR TITLE
add config validation as admission controller

### PR DESCRIPTION
**Pull request by nimakaviani**
This PR implements the proposal discussed in https://github.com/knative/serving/issues/4610

It adds a validating webhook on ConfigMaps in a knative component namespace.

### CODE REVIEW

**Comment by vagababov** [comment]
/ok-to-test

@nimakaviani could you actually fill in the description. The title is mostly self-explanatory, but it's good to have more detailed context of the changes within.

File: webhook/config_validation_controller.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+func (ac *ConfigValidationController) Admit(ctx context.Context, request *admissionv1beta1.AdmissionRequest) *admissionv1beta1.AdmissionResponse {
+	logger := logging.FromContext(ctx)
+	switch request.Operation {
+	case admissionv1beta1.Create, admissionv1beta1.Update:
+	default:
+		logger.Infof("Unhandled webhook operation, letting it through %v", request.Operation)
+		return &admissionv1beta1.AdmissionResponse{Allowed: true}
+	}
+
+	errrr := ac.validate(ctx, request)
```
Comment (vagababov) [inline webhook/config_validation_controller.go]: eh, `err`? 

https://imgflip.com/i/39s6cg

Comment (nimakaviani) [reply webhook/config_validation_controller.go]: 😄thats to show how angry the controller becomes. will fix thx!

**Comment by nimakaviani** [comment]
/hold
while merging https://github.com/knative/pkg/pull/653

File: webhook/webhook.go [hunk 2]
```diff
@@ -95,6 +99,13 @@ type ControllerOptions struct {
 	// Service path for ResourceAdmissionController webhook
 	// Default is "/" for backward compatibility and is set by the constructor
 	ResourceAdmissionControllerPath string
+
+	// Service path for ConfigValidationController webhook
+	// Default is "/config-validation" and is set by the constructor
+	ConfigValidationControllerPath string
+
+	// NamespaceLabel is the label for the Namespace we bind ConfigValidationController to
+	NamespaceLabel string
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply webhook/webhook.go]: yeah unfortunately using namespace selectors appears to be the only way 😔

**Comment by nimakaviani** [comment]
/lint

updated based on @mattmoor 's suggestions

File: webhook/config_validation_controller.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+		panic("config constructor must be a function")
+	}
+
+	if cType.NumIn() != 1 || cType.In(0) != reflect.TypeOf(&corev1.ConfigMap{}) {
+		panic("config constructor must be of the type func(*k8s.io/api/core/v1/ConfigMap) (..., error)")
+	}
+
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+
+	if cType.NumOut() != 2 || !cType.Out(1).Implements(errorType) {
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply webhook/config_validation_controller.go]: added the following:

https://github.com/knative/pkg/pull/636/files#diff-ef8b491dc25634825c642082933e7f6aR49-R61

File: configmap/filter.go [hunk 4]
```diff
@@ -42,3 +45,17 @@ func TypeFilter(ts ...interface{}) func(func(string, interface{})) func(string,
 		}
 	}
 }
+
+// ValidateType checks the type of the constructor it evaluates
+// for expected returned values.
+//
+// The expectation is for the constructor to return two values
+// with the second value being of type error
+func ValidateType(cType reflect.Type) error {
```
Comment (dprotaso) [inline configmap/filter.go]: You can probably pull more validation into this method - there's still some duplication.

I'd also rename the method to be more specific ie. `ValidateConstructor`

Comment (mattmoor) [reply configmap/filter.go]: Can we just have it take the constructor itself?

Comment (vaikas-google) [reply configmap/filter.go]: yeah, if you pass the constructor here, you can just check in one place that it's a Func and the expected signature.

Comment (nimakaviani) [reply configmap/filter.go]: yeah overlooked the lines prior. fixed.

File: webhook/config_validation_controller.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+					Key:      ac.options.ConfigValidationNamespaceLabel,
+					Operator: metav1.LabelSelectorOpExists,
+				}},
+			},
+			FailurePolicy: &failurePolicy,
+		}},
+	}
+
+	// Set the owner to our deployment.
+	deployment, err := kubeClient.Apps().Deployments(ac.options.Namespace).Get(ac.options.DeploymentName, metav1.GetOptions{})
```
Comment (mattmoor) [inline webhook/config_validation_controller.go]: Apps goes away in modern clients, so use:
```suggestion
	deployment, err := kubeClient.AppsV1().Deployments(ac.options.Namespace).Get(ac.options.DeploymentName, metav1.GetOptions{})
```

Comment (mattmoor) [reply webhook/config_validation_controller.go]: I cleaned this up in serving yesterday, and need to split off a change to do the rest of pkg today.

Comment (nimakaviani) [reply webhook/config_validation_controller.go]: sounds good. I saw you are bumping k8s lib, so with the new lib we can also scope the rule below to `namespace:

https://github.com/knative/pkg/blob/8d2bd46f52b3df0d05ddf12f8946cf02488d6a80/webhook/config_validation_controller.go#L90-L100

Comment (nimakaviani) [reply webhook/config_validation_controller.go]: i'll wait for your changes to `pkg` to land before rebasing and finalizing.

Comment (mattmoor) [reply webhook/config_validation_controller.go]: That may take a few days, because we need a K8s change as well.

Comment (nimakaviani) [reply webhook/config_validation_controller.go]: ok updated config validation to use `kubeClient.AppsV1()` as you suggested.

Comment (mattmoor) [reply webhook/config_validation_controller.go]: Alright, the K8s lib changes have landed, so we can start to use 1.14 features.

Comment (nimakaviani) [reply webhook/config_validation_controller.go]: done.

**Pull request merged**